project4/name_server/bootstrap.cpp

```diff
@@ -1,40 +1,218 @@
+    client_ = std::make_unique<message_passing::Client>(threadpool_, successor_);
+    client_->Send(entrance_info);
+  }
 
-void Bootstrap::InitiateEntrance(const message_passing::Endpoint server) {}
+}
 
-std::string Bootstrap::LookUp(int key) { return ""; }
+void Bootstrap::HandleRequest(
+    const consistent_hash_msgs::EntranceInformation& request) {
+  consistent_hash_msgs::EntranceInformation req = request;
+  if (req.successor_info().IsInitialized() && !req.predecessor_info().IsInitialized()) {
+    // the entering server will now be the last in the ring.
+    req.mutable_predecessor_info()->set_id(predecessor_id_);
+    req.mutable_predecessor_info()->set_ip(predecessor_.hostname);
+    req.mutable_predecessor_info()->set_port(predecessor_.port);
+  } else if (!request.successor_info().IsInitialized() && request.predecessor_info().IsInitialized()) {
+    // the entering server will now be the first in the ring.
+    req.mutable_successor_info()->set_port(successor_.port);
+    req.mutable_successor_info()->set_ip(successor_.hostname);
+    req.mutable_predecessor_info()->set_id(successor_id_);
+  } else if (!(req.mutable_predecessor_info()->IsInitialized() || req.mutable_successor_info()->IsInitialized())) {
+    LOG_F(FATAL, "EntranceInformation message is uninitialized after returning to the bootstrap.");
+    return;
+  }
+  client_ = std::make_unique<message_passing::Client>(threadpool_, entering_nameserver_);
+  if (!client_->SendAsync(request)) {
+    LOG_F(ERROR, "Request failed to send.");
+    // error
+  }
+}
 
-void Bootstrap::Insert(int key, std::string val) {}
+void Bootstrap::LookUp(uint key) {
+  // key-val resides in this nameserver
+  const auto itr = pairs_.find(key);
+  if (itr != pairs_.end()) {
+    // key-val pair found , now print
+    console_task_->SendConsole(std::string("Value: ").append(itr->second));
+    console_task_->SendConsole("Contacted: 0");
+  } else if (successor_ == bootstrap_) {
+    console_task_->SendConsole("Key not found.");
+  } else {
+    std::cout << "sending lookup" << std::endl;
+    consistent_hash_msgs::LookUpResult lookup;
+    lookup.set_id(0);
+    lookup.set_key(key);
+    client_ =
+        std::make_unique<message_passing::Client>(threadpool_, successor_);
+    if (!client_->SendAsync(lookup)) {
+      // error
+      LOG_F(ERROR, "Request failed to send.");
+    }
+    std::cout << "sent lookup" << std::endl;
+  }
+}
 
-void Bootstrap::Delete(int key, std::string val) {}
+void Bootstrap::Insert(uint key, std::string val) {
+  if (key >= bounds_.first && key <= bounds_.second) {
+    // key-val should be inserted in this bootstrap
+    std::pair<int, std::string> pair(key, val);
+    pairs_.insert(pair);
+    console_task_->SendConsole("Inserted into: 0");
+    console_task_->SendConsole("Contacted: 0");
+  } else if (successor_ == bootstrap_) {
+    console_task_->SendConsole("Key out of range.");
+  } else {
+    consistent_hash_msgs::InsertResult insert;
+    insert.set_id(0);
```

Comment:
I think here you're supposed to be calling `set_value(val)` instead. For `InsertResults`, the `id` is only set by the nameserver that the key is inserted into.

Reply:
I added `set_value()` but I kept `set_id()` because, when the bootstrap sees the message later, if `id` is `0`, the bootstrap knows it was not inserted into any nameserver.